pkg/common/log/reopen.go

```diff
@@ -0,0 +1,91 @@
+package log
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+const (
+	fileFlags = os.O_APPEND | os.O_CREATE | os.O_WRONLY
+	fileMode  = 0640
+)
+
+var _ ReopenableWriteCloser = (*ReopenableFile)(nil)
+
+// Reopener inspired by https://github.com/client9/reopen
+type Reopener interface {
+	Reopen() error
+}
+
+type ReopenableWriteCloser interface {
+	Reopener
+	io.WriteCloser
+}
+
+type ReopenableFile struct {
+	name string
+	f    *os.File
+	mu   sync.Mutex
+}
+
+func NewReopenableFile(name string) (*ReopenableFile, error) {
+	file, err := os.OpenFile(name, fileFlags, fileMode)
+	if err != nil {
+		return nil, err
+	}
+	return &ReopenableFile{
+		name: name,
+		f:    file,
+	}, nil
+}
+
+func (r *ReopenableFile) Reopen() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	newFile, err := os.OpenFile(r.name, fileFlags, fileMode)
+	if err != nil {
+		r.f = nil
+		return err
+	}
+
+	if r.f != nil {
+		if err := r.f.Close(); err != nil {
+			return err
```

Comment:
Do we want to tear down the server if we can't close the file descriptor for the old log file? I wonder if we shouldn't just kick off a goroutine that logs a warning...

Reply:
Again, where would this warning be logged?

Reply:
If we can't close the old log file, it will leak file descriptors. If the close fails, we may still be able to write to it. If we kick off a goroutine, we could compound the fd leak with a goroutine leak.

Reply:
We're waxing a bit theoretical here, but, as far as I understand, under no circumstances is it safe to try and use the file to write with (or even retry the close) after a failure to close. Even if the failure is EINTR, posix says it should remain open, but not all platforms behave that way (e.g. linux). Because of these reasons, Go will "close" the file. Subsequent calls to Write or Close will fail.

Kicking off a goroutine that logs shouldn't result in a leak unless the logging call blocks, which would mean other logging calls are blocking and the server would be in a terrible state already.

In any case, I don't think any of these circumstances really benefits from killing the server. I'm ok if we don't try and log the failure to close, but I still feel that we shouldn't tear down the server if that happens.